### PR TITLE
WIP: ppx

### PR DESCRIPTION
**Pull request by j0sh**
Also added some tests for the format-string parser, although the file is not in oasis yet.

The main test suite `t_sqlexpr_sqlite.ml` has also been converted to ppx, however it doesn't currently compile due to this strange type error:

    josh@vesper:~/ocaml/ocaml-sqlexpr$ ocamlfind ocamlc -package sqlexpr.ppx,lwt.ppx,oUnit -thread -linkpkg -ppxopt lwt.ppx,-no-debug tests/t_ppx_sqlexpr.ml
    File "tests/t_ppx_sqlexpr.ml", line 50, characters 8-70:
    Error: This expression has type unit S.result = unit Lwt/1063.t
       but an expression was expected of type 'a Lwt/0.t

As far as I can tell, this is due to the wizardry in the `Test` module that redefines lwt and sqlexpr. Not sure how to fix this.

Note that because lwt is redefined, we need to disable lwt debug due to lwt's syntax extensions calling some lwt-specific functions (backtrace_catch and friends), which don't exist in the redefined signature.

### CODE REVIEW

File: src/ppx/ppx_sqlexpr.ml [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+  fun () ->
+    incr n;
+    sprintf "__ppx_sql_%d" !n
+
+let gen_stmt loc ~cacheable sql inp =
+  let mkstr s = Exp.constant (Const_string (s, None)) in
+  let mklid s = {txt=Lident s; loc} in
+  let mkident s = Exp.ident (mklid s) in
+  let mkapply fn args = Exp.apply (mkident fn) args in
+  let mkpident txt = Pat.var {txt; loc} in
```
Comment (Drup) [inline src/ppx/ppx_sqlexpr.ml]: Pretty sure half of those are in `st_convenience`.

**Comment by Drup** [comment]
The location are not added for most of the generated parsetree. You should try to create ill-typed expressions and see where the error message points.

(Also, I have bad opinions about starting a module with 6 open, some of them largely optional, but I'm not the maintainer.)

File: src/ppx/ppx_sqlexpr.ml (hunk 1 above)
Comment (j0sh) [inline src/ppx/ppx_sqlexpr.ml]: Fixed, didn't know about the `Ast_convenience` module. Thanks!

**Comment by j0sh** [comment]
Locations should be fixed now (via Ast_helper.with_default_loc) and the opens have been trimmed a bit.

**Comment by michipili** [comment]
CC me

**Comment by mfp** [comment]
I've merged & adapted the OMake-based build system [in my ppx branch](https://github.com/mfp/ocaml-sqlexpr/tree/ppx).

I found the cause of the `... an expression was expected of type 'a Lwt/0.t` type error, [it was the outer `(>>=)`](https://github.com/mfp/ocaml-sqlexpr/commit/4634364911e1257d8363e71143c5b01d22564293).

Now it's choking on this:


        S.select db [%sql "SELECT @d{%d}"] 42 >|= aeq_list ~printer:(sprintf "%d") [42] >>= fun () ->

        File "t_ppx_sqlexpr.ml", line 121, characters 22-25:
        Error: The field Sqlexpr.sql_statement belongs to the record type
                 Sqlexpr.statement
               but a field was expected belonging to the record type
                 S.expression = Sqlexpr.expression

File: src/ppx/sqlexpr_parser.ml [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    let mtch = Re.get substrings 0 in
+    esc_list := mtch :: !esc_list;
+    esc_str in
+
+  let escaped = Re.replace ~f:esc_subst escrgx str in
+  esc_list := List.rev !esc_list;
+
+  (* logic to extract inputs and outputs *)
+  let inrgx = Re_pcre.regexp {|%([dlLfsSba])(\?)?|} in
+  let outrgx = Re_pcre.regexp {|@([dlLfsSba])(\?)?\{((\w|\.)+)\}|} in
```
Comment (mfp) [inline src/ppx/sqlexpr_parser.ml]: This seems to be the culprit of ppx_sqlexpr choking on the likes of `[%sql "SELECT @d{%d}"]`.

Comment (j0sh) [inline src/ppx/sqlexpr_parser.ml]: Indeed. Fixed this and all the other remaining issues, sent another PR that goes into your ppx branch: https://github.com/mfp/ocaml-sqlexpr/pull/10
